src/tqec/circuit/detectors/boundary.py

```diff
@@ -0,0 +1,238 @@
+from __future__ import annotations
+
+import typing as ty
+
+import numpy
+from tqec.circuit.detectors.measurement import RelativeMeasurementLocation
+from tqec.circuit.detectors.pauli import PauliString
+from tqec.exceptions import TQECException
+
+
+class BoundaryStabilizer:
+    def __init__(
+        self,
+        stabilizer: PauliString,
+        collapsing_operations: ty.Iterable[PauliString],
+        measurements: list[RelativeMeasurementLocation],
+        reset_qubits: frozenset[int],
+        forward: bool,
+    ):
+        """Represents a stabilizer that has been propagated and is now at the boundary
+        of a Fragment.
+
+        Raises:
+            TQECException: if `source_qubits` is empty.
+
+        Args:
+            stabilizer: The propagated stabilizer **before** any collapsing operation
+                has been applied.
+            collapsing_operations: The collapsing operations the stabilizer will have to
+                go through to exit the Fragment.
+            involved_measurements: measurement offsets relative to the **end** of the
+                fragment (even if the created BoundaryStabilizer instance represents a
+                stabilizer on the beginning boundary) of measurements that are involved
+                in this stabilizer.
+            source_qubit: index of the qubit on which the collapsing operation that started
+                the propagation of this boundary stabilizer was applied on.
+                Should contain at least one index.
+        """
+        self._stabilizer = stabilizer
+        self._measurements = measurements
+        self._collapsing_operations = collapsing_operations
```

Comment:
Yes, this property is needed in the `after_collapse` property, because the `PauliString.collapse_by` method requires such a pre-condition. Good catch!

Reply:
After a quick look, the pre-condition "provided collapsing operators should not be duplicated" was required per the documentation, but not really required by the code. I added the `fronzenset` because I find that cleaner, and also added documentation to the `PauliString.collapse_by` method.